In [ ]:
from glob import glob
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from keras.src.layers import Rescaling
from tensorflow.keras.layers import Dense, Dropout, Flatten, Conv2D
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential
from keras import Input
from keras.src.optimizers import Adam
from sklearn.metrics import ConfusionMatrixDisplay, classification_report
from keras.src.metrics.metrics_utils import confusion_matrix
from sklearn.metrics import precision_score, recall_score


fashion_path_train = glob("../Fashion_MNIST/fashion-mnist_train.csv")
fashion_path_test = glob("../Fashion_MNIST/fashion-mnist_test.csv")
# fish_path
fashion_data_train_ = pd.read_csv(fashion_path_train[0])
fashion_data_test_ = pd.read_csv(fashion_path_test[0])

# fashion_data_train_.shape
# fashion_data_test_.shape

# fashion_data_train_.isnull().sum().sum()
# fashion_data_train_.isna().sum().sum()
# fashion_data_test_.isnull().sum().sum()
# fashion_data_test_.isna().sum().sum()

### Preprocessing

In [ ]:
# from sklearn.preprocessing import StandardScaler

X_train = fashion_data_train_.drop(['label'], axis = 1)
X_test = fashion_data_test_.drop(['label'], axis = 1)
y_train = fashion_data_train_["label"]
y_test = fashion_data_test_["label"]

y_train_cat = keras.utils.to_categorical(y_train, num_classes=10)
y_test_cat  = keras.utils.to_categorical(y_test,  num_classes=10)

### Model Dense, Compile, Train, Predict

In [31]:
model = Sequential([
    Input(shape=(784,)),
    Rescaling(1./255),
    Dense(64, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])


model.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy')


model.fit(X_train, y_train_cat, epochs=30, batch_size=32, validation_split=0.1)

model.evaluate(X_test, y_test_cat)

y_predi = model.predict(X_test)

y_pred = np.argmax(y_predi, axis=1)

y_pred

Epoch 1/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 2s 439us/step - loss: 0.5313 - val_loss: 0.4058
Epoch 2/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 391us/step - loss: 0.3855 - val_loss: 0.3756
Epoch 3/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 385us/step - loss: 0.3483 - val_loss: 0.3739
Epoch 4/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 385us/step - loss: 0.3270 - val_loss: 0.3493
Epoch 5/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 386us/step - loss: 0.3110 - val_loss: 0.3502
Epoch 6/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 386us/step - loss: 0.2998 - val_loss: 0.3387
Epoch 7/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 385us/step - loss: 0.2884 - val_loss: 0.3473
Epoch 8/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 390us/step - loss: 0.2763 - val_loss: 0.3527
Epoch 9/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 390us/step - loss: 0.2698 - val_loss: 0.3280
Epoch 10/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 383us/step - loss: 0.2607 - val_loss: 0.3379
Epoch 11/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 384us/step - loss: 0.2528 - val_loss: 0.3565
Epoch 12

array([0, 1, 2, ..., 8, 8, 1], shape=(10000,))

In [ ]:
print(y_pred)
print(y_test)

In [32]:
confusion_m = confusion_matrix(y_test, y_pred, num_classes=10)

precision = precision_score(y_test, y_pred, average="macro")
recall    = recall_score(y_test, y_pred, average="macro")

print("Precision :", precision)
print("Recall    :", recall)


print(f" CONFUSION MATRIX : \n {confusion_m}")
print(f" \n CLASSIFICATION REPORT : \n {classification_report(y_test, y_pred)}")


Precision : 0.8909301083746437
Recall    : 0.8897
 CONFUSION MATRIX : 
 [[876   2  12  14   1   0  85   0  10   0]
 [  5 987   0   7   0   1   0   0   0   0]
 [ 27   1 769  14 130   0  58   0   1   0]
 [ 35  14   5 897  33   0  14   0   2   0]
 [  2   1  49  21 898   0  27   0   2   0]
 [  2   0   0   1   0 956   0  27   4  10]
 [172   2  58  24  93   0 643   0   8   0]
 [  0   0   0   0   0  17   0 961   0  22]
 [  5   0   3   3   6   1   9   1 972   0]
 [  1   0   0   0   0   3   0  57   1 938]]
 
 CLASSIFICATION REPORT : 
               precision    recall  f1-score   support

           0       0.78      0.88      0.82      1000
           1       0.98      0.99      0.98      1000
           2       0.86      0.77      0.81      1000
           3       0.91      0.90      0.91      1000
           4       0.77      0.90      0.83      1000
           5       0.98      0.96      0.97      1000
           6       0.77      0.64      0.70      1000
           7       0.92      0.96  

## Test de plusieurs paramètres

### Model A

In [24]:
epochs = [30, 40, 50, 60, 70]

neurons = [32, 64, 128, 256, 512]

In [25]:
precisions_A = {}
all_precisions_A = {}

for neuron in neurons:
    for epoch in epochs:
        model_A = Sequential([
            Input(shape=(784,)),
            Rescaling(1./255),
            Dense(neuron, activation="relu"),
            Dense(neuron, activation="relu"),
            Dense(10, activation="softmax")
        ])

        model_A.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy')

        model_A.fit(X_train, y_train_cat, epochs=epoch, batch_size=32, validation_split=0.1)

        model_A.evaluate(X_test, y_test_cat)

        y_predi_A = model_A.predict(X_test)

        y_pred_A = np.argmax(y_predi_A, axis=1)

        precisions_A[epoch] = precision_score(y_test, y_pred_A, average="macro")

    all_precisions_A[neuron] = precisions_A
    precisions_A = {}

Epoch 1/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 366us/step - loss: 0.5638 - val_loss: 0.4790
Epoch 2/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 324us/step - loss: 0.4136 - val_loss: 0.4049
Epoch 3/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 325us/step - loss: 0.3800 - val_loss: 0.3843
Epoch 4/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 323us/step - loss: 0.3578 - val_loss: 0.4279
Epoch 5/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 321us/step - loss: 0.3425 - val_loss: 0.3726
Epoch 6/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 321us/step - loss: 0.3269 - val_loss: 0.3601
Epoch 7/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 327us/step - loss: 0.3185 - val_loss: 0.3602
Epoch 8/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 321us/step - loss: 0.3084 - val_loss: 0.3843
Epoch 9/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 322us/step - loss: 0.3017 - val_loss: 0.3438
Epoch 10/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 321us/step - loss: 0.2948 - val_loss: 0.3432
Epoch 11/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.2905 - val_loss: 0.3477
Epoch 12

In [28]:
all_precisions_A_df = pd.DataFrame(all_precisions_A)
all_precisions_A_df

,32,64,128,256,512
30,0.882889,0.891513,0.891722,0.891108,0.897644
40,0.879738,0.892693,0.891840,0.898835,0.896746
50,0.876266,0.895133,0.893676,0.892106,0.896988
60,0.877802,0.884103,0.895776,0.893904,0.897561
70,0.885523,0.886207,0.895736,0.898479,0.889075


##### POur ce model les meilleurs paramètres sont : nbre de neurones = 256 et epochs = 40. La précision est de 0.898835

### Model B

In [26]:
precisions_B = {}
all_precisions_B = {}

for neuron in neurons:
    for epoch in epochs:
        model_B = Sequential([
            Input(shape=(784,)),
            Rescaling(1./255),
            Dense(neuron, activation="relu"),
            Dense(neuron, activation="relu"),
            Dense(neuron, activation="relu"),
            Dense(10, activation="softmax")
        ])

        model_B.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy')

        model_B.fit(X_train, y_train_cat, epochs=epoch, batch_size=32, validation_split=0.1)

        model_B.evaluate(X_test, y_test_cat)

        y_predi_B = model_B.predict(X_test)

        y_pred_B = np.argmax(y_predi_B, axis=1)

        precisions_B[epoch] = precision_score(y_test, y_pred_B, average="macro")

    all_precisions_B[neuron] = precisions_B
    precisions_B = {}




Epoch 1/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 361us/step - loss: 0.5942 - val_loss: 0.4762
Epoch 2/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 319us/step - loss: 0.4233 - val_loss: 0.4119
Epoch 3/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 319us/step - loss: 0.3836 - val_loss: 0.4026
Epoch 4/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 319us/step - loss: 0.3616 - val_loss: 0.3894
Epoch 5/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 321us/step - loss: 0.3433 - val_loss: 0.3707
Epoch 6/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 319us/step - loss: 0.3333 - val_loss: 0.3688
Epoch 7/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.3172 - val_loss: 0.3653
Epoch 8/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.3114 - val_loss: 0.3763
Epoch 9/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.3043 - val_loss: 0.3962
Epoch 10/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.2971 - val_loss: 0.3667
Epoch 11/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 318us/step - loss: 0.2878 - val_loss: 0.3835
Epoch 12

In [29]:
all_precisions_B_df = pd.DataFrame(all_precisions_B)
all_precisions_B_df

,32,64,128,256,512
30,0.888070,0.891768,0.895320,0.897350,0.887000
40,0.879409,0.887228,0.898039,0.897341,0.900126
50,0.885017,0.894382,0.894854,0.901746,0.895220
60,0.877356,0.889529,0.894407,0.895990,0.897162
70,0.878638,0.887385,0.895871,0.894091,0.900960


##### POur ce model les meilleurs paramètres sont : nbre de neurones = 256 et epochs = 50. La précision est de 0.901746

### Model C

In [27]:
precisions_C = {}
all_precisions_C = {}

for neuron in neurons:
    for epoch in epochs:
        model_C = Sequential([
            Input(shape=(784,)),
            Rescaling(1./255),
            Dense(neuron, activation="relu"),
            Dense(10, activation="softmax")
        ])

        model_C.compile(optimizer=Adam(learning_rate=0.001), loss='categorical_crossentropy')

        model_C.fit(X_train, y_train_cat, epochs=epoch, batch_size=32, validation_split=0.1)

        model_C.evaluate(X_test, y_test_cat)

        y_predi_C = model_C.predict(X_test)

        y_pred_C = np.argmax(y_predi_C, axis=1)

        precisions_C[epoch] = precision_score(y_test, y_pred_C, average="macro")

    all_precisions_C[neuron] = precisions_C
    precisions_C = {}

Epoch 1/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 348us/step - loss: 0.5717 - val_loss: 0.4709
Epoch 2/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.4304 - val_loss: 0.4287
Epoch 3/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 299us/step - loss: 0.3964 - val_loss: 0.4102
Epoch 4/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.3763 - val_loss: 0.4404
Epoch 5/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.3591 - val_loss: 0.3933
Epoch 6/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 302us/step - loss: 0.3462 - val_loss: 0.3852
Epoch 7/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.3350 - val_loss: 0.3849
Epoch 8/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.3260 - val_loss: 0.3868
Epoch 9/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 301us/step - loss: 0.3185 - val_loss: 0.3876
Epoch 10/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 300us/step - loss: 0.3132 - val_loss: 0.3876
Epoch 11/30
1688/1688 ━━━━━━━━━━━━━━━━━━━━ 1s 299us/step - loss: 0.3065 - val_loss: 0.3687
Epoch 12

In [30]:
all_precisions_C_df = pd.DataFrame(all_precisions_C)
all_precisions_C_df

,32,64,128,256,512
30,0.880956,0.887974,0.892689,0.895330,0.896895
40,0.882386,0.892490,0.897317,0.900927,0.897667
50,0.881086,0.884599,0.885706,0.895796,0.901690
60,0.884654,0.885383,0.893227,0.899019,0.892509
70,0.879187,0.879617,0.894920,0.896541,0.898153


##### POur ce model les meilleurs paramètres sont : nbre de neurones = 512 et epochs = 50. La précision est de 0.901690